### PR TITLE
V1.11.x: Status reporting Scale Fixes

### PR DESCRIPTION
**Pull request by ben-taussig-solo**
# Description

- Backports a series of updates made in master (currently v1.13.0-beta) and v1.12.x in order to improve status reporting performance
  - In particular, we backport https://github.com/solo-io/gloo/pull/7145 and https://github.com/solo-io/gloo/pull/7183
- My testing suggests that these changes have drastically improved status reporting performance with large numbers of RouteTables
  - Before this change, I measured 237.6 seconds to update VS status for a VS with 500 RouteTables
  - After the change, I measured 10.8 seconds to update VS status for a VS with 500 RouteTables
- We need to be cautious about whether this work should be merged and released to the main v1.11.x branch or if a new stub branch off of v1.11.x (`v1.11.x-status-patch` ?) should be created, because upgrading an existing cluster to use these changes requires a somewhat painful CRD upgrade
  - The PR currently targets v1.11.x, although I will leave it as a draft until we have decided how to proceed

BOT NOTES: 
resolves https://github.com/solo-io/solo-projects/issues/4076

### CODE REVIEW

File: projects/gateway/pkg/syncer/translator_syncer.go [hunk 1]
```diff
@@ -411,13 +411,23 @@ func (s *statusSyncer) syncStatus(ctx context.Context) error {
...
+		// while tempting to write statuses in parallel to increase performance, we should actually first consider recommending the user tunes k8s qps/burst:
+		// https://github.com/solo-io/gloo/blob/a083522af0a4ce22f4d2adf3a02470f782d5a865/projects/gloo/api/v1/settings.proto#L337-L350
 		if err := s.reporter.WriteReports(ctx, reports, currentStatuses); err != nil {
-			errs = multierror.Append(errs, err)
+			// add TEMPORARY wrap to our WriteReports error that we should remove in Gloo Edge ~v1.16.0+.
+			// to get the status performance improvements, we need to make the assumption that the user has the latest CRDs installed.
+			// if a user forgets the error message is very confusing (invalid request during kubectl patch);
+			// this should help them understand what's going on in case they did not read the changelog.
+			wrappedErr := errors.Wrapf(err, "failed to write reports for %v;"+
+				"did you make sure your CRDs have been updated since v1.12.17? (i.e. `status` and `status.statuses` fields exist on your CR)", inputResource.GetMetadata().Ref().Key())
```
Comment (kdorosh) [inline projects/gateway/pkg/syncer/translator_syncer.go:423]: update the message here to be the correct 1.11 version

File: projects/gloo/pkg/syncer/translator_syncer.go [hunk 2]
```diff
@@ -106,10 +106,20 @@ func (s *translatorSyncer) Sync(ctx context.Context, snap *v1snap.ApiSnapshot) e
...
+	//
+	// add TEMPORARY wrap to our WriteReports error that we should remove in Gloo Edge ~v1.16.0+.
+	// to get the status performance improvements, we need to make the assumption that the user has the latest CRDs installed.
+	// if a user forgets the error message is very confusing (invalid request during kubectl patch);
+	// this should help them understand what's going on in case they did not read the changelog.
 	if err := s.reporter.WriteReports(ctx, reports, nil); err != nil {
 		logger.Debugf("Failed writing report for proxies: %v", err)
-		multiErr = multierror.Append(multiErr, eris.Wrapf(err, "writing reports"))
+		wrappedErr := eris.Wrapf(err, "failed to write reports"+
+			"did you make sure your CRDs have been updated since v1.12.17 of open-source? (i.e. `status` and `status.statuses` fields exist on your CR)")
```
Comment (kdorosh) [inline projects/gloo/pkg/syncer/translator_syncer.go:119]: update the message here to be the correct 1.11 version

File: projects/ingress/pkg/api/service/resource_client.go [hunk 3]
```diff
@@ -170,6 +172,36 @@ func (rc *ResourceClient) Write(resource resources.Resource, opts clients.WriteO
...
 
+func (rc *ResourceClient) ApplyStatus(statusClient resources.StatusClient, inputResource resources.InputResource, opts clients.ApplyStatusOpts) (resources.Resource, error) {
+	name := inputResource.GetMetadata().GetName()
+	namespace := inputResource.GetMetadata().GetNamespace()
+	if err := resources.ValidateName(name); err != nil {
+		return nil, errors.Wrapf(err, "validation error")
+	}
+	opts = opts.WithDefaults()
+
+	data, err := shared.GetJsonPatchData(context.TODO(), inputResource)
```
Comment (kdorosh) [inline projects/ingress/pkg/api/service/resource_client.go:183]: context todo? shouldn't this be opts.Ctx?

File: projects/knative/pkg/api/custom/knative/knative_clusteringress_client.go [hunk 4]
```diff
@@ -75,6 +75,10 @@ func (rc *ResourceClient) Delete(namespace, name string, opts clients.DeleteOpts
 	panic("this client does not support delete operations")
 }
 
+func (rc *ResourceClient) ApplyStatus(statusClient resources.StatusClient, inputResource resources.InputResource, opts clients.ApplyStatusOpts) (resources.Resource, error) {
+	panic("this client does not support apply status operations")
```
Comment (kdorosh) [inline projects/knative/pkg/api/custom/knative/knative_clusteringress_client.go:79]: i slightly prefer changing all the new panics in this PR to return errors to be extra safe about the backport here

**Review by kdorosh — COMMENTED** [review]
generally lgtm. CRD updates for patch releases are painful, but the risk of not doing so only affects status reporting (DP traffic is fine and still updates) so it's mostly a UX thing and one-time upgrade step to get large benefits. I feel ok bringing this set of changes to 1.11 main branch

File: projects/ingress/pkg/api/service/resource_client.go (hunk 3 above)
Comment (ben-taussig-solo) [reply projects/ingress/pkg/api/service/resource_client.go]: Resolved here: https://github.com/solo-io/gloo/pull/7347/commits/e8d374bbec776ca7cbec01f62776767042f87ab2

File: projects/gloo/pkg/syncer/translator_syncer.go (hunk 2 above)
Comment (ben-taussig-solo) [reply projects/gloo/pkg/syncer/translator_syncer.go]: Resolved here: https://github.com/solo-io/gloo/pull/7347/commits/e8d374bbec776ca7cbec01f62776767042f87ab2

File: projects/gateway/pkg/syncer/translator_syncer.go (hunk 1 above)
Comment (ben-taussig-solo) [reply projects/gateway/pkg/syncer/translator_syncer.go]: Resolved here: https://github.com/solo-io/gloo/pull/7347/commits/e8d374bbec776ca7cbec01f62776767042f87ab2

**Review by nfuden — APPROVED** [review]

**Pull request merged**